File: day09/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <sstream>
#include  <regex>
#include <optional>
#include <functional>
#include <set>
#include <unordered_map>
#include <map>

std::vector<int64_t> find_integers(const std::string &str) {
    static std::regex base_regex("(-?\\d+)");

    std::vector<int64_t> result;
    for (auto it = std::sregex_iterator(str.begin(), str.end(), base_regex);
         it != std::sregex_iterator(); ++it) {

        auto value = std::stoll(it->operator[](1).str());
        result.push_back(value);
    }
    return result;
}

int64_t solve(bool backwards) {

    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");

    int64_t result = 0;
    std::string line;
    while(std::getline(ifs, line))
    {
        std::vector<std::vector<int64_t>> history = {find_integers(line)};
        while(!std::all_of(history.back().begin(), history.back().end(), [](int64_t value){ return value==0;})) {
            const auto& last = history.back();
            std::vector<int64_t> newline;
            for(int i=1; i < last.size(); ++i) {
                newline.push_back(last[i]-last[i-1]);
            }
            history.emplace_back(std::move(newline));
        }

        int64_t predicted = 0;
        for(int i=history.size()-2; i >= 0; --i) {
            if(backwards)
                predicted = history[i].front()-predicted;
            else
                predicted = history[i].back()+predicted;
        }
        result+=predicted;
    }

    return result;
}


int main() {

    std::cout << "Part1: " << solve(false) << std::endl;
    std::cout << "Part2: " << solve(true) << std::endl;
    return 0;
}

File: day09/src/main.rs
use std::collections::{HashMap, HashSet};
use lazy_static::lazy_static;
use regex::{Match, Regex};
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;
use std::vec;
use gcd::binary_u64;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let path = Path::new(file!());
    let Ok(lines) = read_lines("input.txt") else {
        panic!("file is not found");
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn find_integers(input: &str) -> Vec<i64> {

    lazy_static! {
        static ref NUMBER_RE: Regex = Regex::new(r#"(?P<number>-?[0-9]+)"#).unwrap();
    }

    let mut result: Vec<i64> = vec![];
    for capture in NUMBER_RE.captures_iter(input) {
        let number = capture.name("number").unwrap().as_str().parse::<i64>().unwrap();
        result.push(number);
    }
    result
}


fn solve(backwards: bool) -> i64 {

    let mut result = 0i64;
    for line in read() {
        let mut history = vec![find_integers(&line)];

        while !history.last().unwrap().iter().all(|&x| x==0) {
            let last = history.last().unwrap();
            let mut newstep: Vec<i64> = vec![];
            for i in 1..last.len() {
                newstep.push(last[i]-last[i-1]);
            }
            history.push(newstep);
        }

        let mut prediction = 0i64;
        for i in (0..history.len()-1).rev() {
            if backwards {
                prediction = history[i].first().unwrap() - prediction;
            } else {
                prediction = history[i].last().unwrap() + prediction;
            }
        }
        result+=prediction;
    }
    result
}

fn main() {

    println!("part1 {}", solve(false));
    println!("part2 {}", solve(true));
}
